drivers/adc/adc_nrfx_saadc.c

```diff
@@ -0,0 +1,374 @@
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		SYS_LOG_ERR("Provided buffer is too small (%u/%u)",
+			    sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int start_read(struct device *dev, const struct adc_sequence *sequence)
+{
+	int error = 0;
+	u32_t channels = sequence->channels;
+	u8_t active_channels = 0;
+	u8_t channel_id;
+
+	/* Signal an error if channel selection is invalid (no channels or
+	 * a non-existing one is selected).
+	 */
+	if (!channels || (channels & (~0UL << NRF_SAADC_CHANNEL_COUNT))) {
+		SYS_LOG_ERR("Invalid selection of channels");
+		return -EINVAL;
+	}
+
+	active_channels = 0;
+
+	/* Enable only the channels selected for the pointed sequence.
+	 * Disable all the rest.
+	 */
+	channel_id = 0;
+	do {
+		if (channels & 1) {
+			/* Signal an error if a selected channel has not been
+			 * configured yet.
+			 */
+			if (m_positive_inputs[channel_id] == 0) {
+				SYS_LOG_ERR("Channel %u not configured",
+					    channel_id);
+				return -EINVAL;
+			}
+			nrf_saadc_channel_pos_input_set(
+				channel_id,
+				m_positive_inputs[channel_id]);
+			++active_channels;
+		} else {
+			nrf_saadc_channel_pos_input_set(
+				channel_id,
+				NRF_SAADC_INPUT_DISABLED);
+		}
+		channels >>= 1;
+	} while (++channel_id < NRF_SAADC_CHANNEL_COUNT);
+
+	error = set_resolution(sequence);
+	if (error) {
+		return error;
+	}
+
+	error = set_oversampling(sequence, active_channels);
+	if (error) {
+		return error;
+	}
+
+	error = check_buffer_size(sequence, active_channels);
+	if (error) {
+		return error;
+	}
+
+	nrf_saadc_buffer_init((nrf_saadc_value_t *)sequence->buffer,
+			      active_channels);
+
+	adc_context_start_read(&m_data.ctx, sequence);
+
+	error = adc_context_wait_for_completion(&m_data.ctx);
+	adc_context_release(&m_data.ctx, error);
+
+	return error;
+}
+
+static int adc_nrfx_read(struct device *dev,
+			 const struct adc_sequence *sequence)
+{
+	adc_context_lock(&m_data.ctx, false, NULL);
+	return start_read(dev, sequence);
+}
+
+#ifdef CONFIG_ADC_ASYNC
+static int adc_nrfx_read_async(struct device *dev,
+			       const struct adc_sequence *sequence,
+			       struct k_poll_signal *async)
+{
+	adc_context_lock(&m_data.ctx, true, async);
+	return start_read(dev, sequence);
+}
+#endif
+
+static void saadc_irq_handler(void *param)
+{
+	struct device *dev = (struct device *)param;
+
+	if (nrf_saadc_event_check(NRF_SAADC_EVENT_END)) {
+		nrf_saadc_event_clear(NRF_SAADC_EVENT_END);
+
+		nrf_saadc_task_trigger(NRF_SAADC_TASK_STOP);
+		nrf_saadc_disable();
+
+		adc_context_on_sampling_done(&m_data.ctx, dev);
+	}
+}
+
+DEVICE_DECLARE(adc_0);
+
+static int init_saadc(struct device *dev)
+{
+	nrf_saadc_event_clear(NRF_SAADC_EVENT_END);
+	nrf_saadc_int_enable(NRF_SAADC_INT_END);
+	NRFX_IRQ_ENABLE(SAADC_IRQn);
+
+	IRQ_CONNECT(SAADC_IRQn, CONFIG_ADC_0_IRQ_PRI,
```

Comment:
IRQ num and priority should come from DT.

Reply:
done